llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp

```diff
@@ -48,6 +49,13 @@
 //   is supported.
 // - Import phase: (same as with hybrid case above).
 //
+// In non-LTO mode:
+// - The pass apply speculative devirtualization without requiring any type of
+// visibility.
+// - Skips other features like virtual constant propagation, uniform return
```

Comment:
It's more fundamental in that I don't think some of these can be applied safely without whole program and in fact without some fullLTO. There is vtable rewriting that happens.